File: client-scalaz/src/main/scala/org/openapitools/client/api/SubscriptionPlan.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import SubscriptionPlan._

case class SubscriptionPlan (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
name: Option[String],
description: Option[String],
options: Option[List[SubscriptionOption]],
price: Option[Double],
promo: Option[Double],
transactionFee: Option[Double],
role: Option[String],
defaultPlan: Option[Boolean],
visible: Option[Boolean],
image: Option[Asset])

object SubscriptionPlan {
  import DateTimeCodecs._

  implicit val SubscriptionPlanCodecJson: CodecJson[SubscriptionPlan] = CodecJson.derive[SubscriptionPlan]
  implicit val SubscriptionPlanDecoder: EntityDecoder[SubscriptionPlan] = jsonOf[SubscriptionPlan]
  implicit val SubscriptionPlanEncoder: EntityEncoder[SubscriptionPlan] = jsonEncoderOf[SubscriptionPlan]
}

File: client-clojure/src/sirqul_io_t_platform/specs/subscription_plan.clj
(ns sirqul-io-t-platform.specs.subscription-plan
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.subscription-option :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            )
  (:import (java.io File)))


(def subscription-plan-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :name) string?
   (ds/opt :description) string?
   (ds/opt :options) (s/coll-of subscription-option-spec)
   (ds/opt :price) float?
   (ds/opt :promo) float?
   (ds/opt :transactionFee) float?
   (ds/opt :role) string?
   (ds/opt :defaultPlan) boolean?
   (ds/opt :visible) boolean?
   (ds/opt :image) asset-spec
   })

(def subscription-plan-spec
  (ds/spec
    {:name ::subscription-plan
     :spec subscription-plan-data}))
